src/Servers/Kestrel/Core/src/Internal/Http/PathNormalizer.cs

```diff
@@ -50,193 +52,144 @@ public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTar
-                    ch2 = *(src + 1);
-
-                    if (ch2 == ByteDot)
-                    {
-                        // B.  if the input buffer begins with a prefix of "/./" or "/.",
-                        //     where "." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer; otherwise,
-                        src += 1;
-                        *src = ByteSlash;
-                        continue;
-                    }
-
-                    break;
-                case 3:
-                    ch2 = *(src + 1);
-                    ch3 = *(src + 2);
-
-                    if (ch2 == ByteDot && ch3 == ByteDot)
-                    {
-                        // C.  if the input buffer begins with a prefix of "/../" or "/..",
-                        //     where ".." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer and remove the last
-                        //     segment and its preceding "/" (if any) from the output
-                        //     buffer; otherwise,
-                        src += 2;
-                        *src = ByteSlash;
-
-                        if (dst > start)
-                        {
-                            do
-                            {
-                                dst--;
-                            } while (dst > start && *dst != ByteSlash);
-                        }
-
-                        continue;
-                    }
-                    else if (ch2 == ByteDot && ch3 == ByteSlash)
-                    {
-                        // B.  if the input buffer begins with a prefix of "/./" or "/.",
-                        //     where "." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer; otherwise,
-                        src += 2;
-                        continue;
-                    }
-
-                    break;
-                default:
-                    ch2 = *(src + 1);
-                    ch3 = *(src + 2);
-                    ch4 = *(src + 3);
+                return src.Length;
+            }
+            if (nextDotSegmentIndex < 0)
+            {
+                // Copy the remianing src to dst, and return.
+                src[readPointer..].CopyTo(src[writtenLength..]);
+                writtenLength += src.Length - readPointer;
+                return writtenLength;
+            }
+            else if (nextDotSegmentIndex > 0)
+            {
+                // Copy until the next segment excluding the trailer.
+                Unsafe.CopyBlock(ref src[writtenLength], ref src[readPointer], (uint)nextDotSegmentIndex);
+                writtenLength += nextDotSegmentIndex;
+                readPointer += nextDotSegmentIndex;
+            }
 
-                    if (ch2 == ByteDot && ch3 == ByteDot && ch4 == ByteSlash)
+            // Case of /../ or /./
```

Comment:
Should this comment be moved 1 line down?
Also, it handles the `/.a` case right?

Reply:
ack